ion/src/test/java/com/fasterxml/jackson/dataformat/ion/polymorphism/PolymorphicRoundTripWithSerializationAnnotations.java

```diff
@@ -0,0 +1,94 @@
+
+import com.amazon.ion.IonValue;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.dataformat.ion.IonGenerator;
+import com.fasterxml.jackson.dataformat.ion.IonObjectMapper;
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class PolymorphicRoundTripWithSerializationAnnotations {
+
+    private static final String SUBCLASS_TYPE_NAME =
+            PolymorphicRoundTripWithSerializationAnnotations.Subclass.class.getTypeName();
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
+    static public abstract class BaseClass { /* empty */ }
+
+    public static class Subclass extends BaseClass {
+        public String someString;
+        public int anInt;
+
+        public Subclass() {};
+        public Subclass(String s, int i) {
+            this.someString = s;
+            this.anInt = i;
+        }
+    }
+
+    @Test
+    public void testNativeTypeIdsEnabledOnWriteByDefault() throws IOException {
+        IonObjectMapper mapper = new IonObjectMapper();
+        Subclass subclass = new Subclass("some value", 42);
+        IonValue subclassAsIon = mapper.writeValueAsIonValue(subclass);
+        String expectedIonStr = "\n" +
+                "'" + SUBCLASS_TYPE_NAME + "'::{\n" +
+                "  someString:\"some value\",\n" +
+                "  anInt:42\n" +
+                "}";
+        assertEquals(expectedIonStr, subclassAsIon.toPrettyString());
+        BaseClass roundTripInstance = mapper.readValue(subclassAsIon, BaseClass.class);
+
+        assertTrue(roundTripInstance instanceof Subclass);
+        assertEquals(subclass.someString, ((Subclass) roundTripInstance).someString);
+        assertEquals(subclass.anInt, ((Subclass) roundTripInstance).anInt);
+    }
+
+    @Test
+    public void testNativeTypeIdsCanBeDisabledOnWrite() throws IOException {
+        IonObjectMapper mapper = new IonObjectMapper()
+                .disable(IonGenerator.Feature.USE_NATIVE_TYPE_ID);
+
+        Subclass subclass = new Subclass("some value", 42);
+        IonValue subclassAsIon = mapper.writeValueAsIonValue(subclass);
+        String expectedIonStr = "\n" +
+                "{\n" +
+                "  '@class':\"" + SUBCLASS_TYPE_NAME + "\",\n" +
+                "  someString:\"some value\",\n" +
+                "  anInt:42\n" +
+                "}";
+        assertEquals(expectedIonStr, subclassAsIon.toPrettyString());
+        BaseClass roundTripInstance = mapper.readValue(subclassAsIon, BaseClass.class);
+
+        assertTrue(roundTripInstance instanceof Subclass);
+        assertEquals(subclass.someString, ((Subclass) roundTripInstance).someString);
+        assertEquals(subclass.anInt, ((Subclass) roundTripInstance).anInt);
+    }
+
+    @Test
+    public void testNativeTypeIdsDisabledOnWriteStillReadsSuccessfully() throws IOException {
```

Comment:
Should this be `testNativeTypeIdsDisabledOn*Read*StillReadsSuccessfully`? The feature is enabled on write and disabled on read.

Reply:
Yes, that might be a better way to phrase it... what I was thinking was that the toggle only actually effects write behavior, it never impacts reads.